src/Acuminator/Acuminator.Analyzers/StaticAnalysis/DacFieldAndReferencedFieldMismatch/DacFieldAndReferencedFieldMismatchAnalyzer.cs

```diff
@@ -0,0 +1,235 @@
+																	.Where(x => x.Item2 is not null && x.Item2 != foreignFieldSize);
+
+				foreach (var (attribute, size) in attributesWithSizeMismatch)
+				{
+					ReportTypeSizeMismatch(
+						context, pxContext, attribute, property.Name,
+						foreignProperty.Name, foreignFieldSize.Value);
+				}
+			}
+		}
+
+		private static bool PropertyTypesAreIncompatible(DacPropertyInfo property)
+		{
+			var attributesWithFieldTypeMetadata = property.Attributes
+														  .Where(aInfo => !aInfo.AggregatedAttributeMetadata.IsDefaultOrEmpty)
+														  .ToList(capacity: property.Attributes.Length);
+			var (typeAttributesOnDacProperty, typeAttributesWithDifferentDataTypesOnAggregator, _) =
+				attributesWithFieldTypeMetadata.FilterTypeAttributes();
+
+			if (typeAttributesWithDifferentDataTypesOnAggregator?.Count > 0)
+				return false;
+
+			var compatibility = property.CheckCompatibility(typeAttributesOnDacProperty![0]);
+
+			return compatibility != TypesCompatibility.CompatibleTypes;
+		}
+
+		private DacSemanticModel? GetOrInferModel(INamedTypeSymbol type, PXContext pxContext, CancellationToken cancellation)
+		{
+			return _modelsCache.TryGetValue(type, out var model)
+				? model
+				: _modelsCache[type] = DacSemanticModel.InferModel(pxContext, type, cancellation: cancellation);
+		}
+
+		private static bool HaveDifferentUnderlyingType(DacPropertyInfo localProperty, ITypeSymbol foreignPropertyType) =>
+			!SymbolEqualityComparer.Default.Equals(localProperty.PropertyType, foreignPropertyType);
+
+		private static INamedTypeSymbol? ExtractForeignFieldSymbol(DacFieldAttributeInfo referenceAttribute, PXContext pxContext)
+		{
+			var argument = referenceAttribute.AttributeData.ConstructorArguments.FirstOrDefault();
+			INamedTypeSymbol? result;
+
+			if (argument.Value is INamedTypeSymbol { IsGenericType: true } symbol)
+				result = ExtractForeignFieldSymbolFromBqlSearch(symbol, pxContext);
+			else
+				result = argument.Value as INamedTypeSymbol;
+
+			return result?.ContainingSymbol is INamedTypeSymbol ? result : null;
+		}
+
+		private static INamedTypeSymbol? ExtractForeignFieldSymbolFromBqlSearch(INamedTypeSymbol searchType, PXContext pxContext)
+		{
+			return searchType.GetBaseTypesAndThis()
+							 .TakeWhile(x => x.SpecialType != SpecialType.System_Object)
+							 .OfType<INamedTypeSymbol>()
+							 .Reverse()
+							 .Where(x => x.IsGenericType && x.ImplementsInterface(pxContext.BQL.IBqlSearch))
+							 .Select(type => type.TypeArguments.FirstOrDefault(arg => arg.IsDacBqlField(pxContext)))
+							 .OfType<INamedTypeSymbol>()
+							 .FirstOrDefault();
+		}
+
+		private static int? GetFieldSize(DacFieldAttributeInfo attrInfo, PXContext pxContext)
+		{
+			if (ContainsAttributeMatchingOneOf(attrInfo, pxContext.FieldAttributes.DataAttributesWithLength)
+				&& GetLengthConstructorArgument(attrInfo.AttributeData, pxContext) is int i)
+				return i;
+
+			var lengthCtorArguments = attrInfo.FlattenedAcumaticaAttributes
```

Comment:
I think, this statement should cover the first `if` case too, because flattened attributes set should contain the attribute itself, making the first `if` redundant.

This code now correctly iterates over the aggregated attributes. But there are still two problems:
- The search for constructor parameter here doesn't use `GetLengthConstructorArgument`
- There is no handling of data type attributes with hard-coded lengths passed to the base constructor like `PXDBEmailAttribute`. 
We previously spoke about creating a special map for such attributes.

Reply:
did as you proposed, but there's one argument - PX.SM.Email.SMDBRecipientAttribute - which have length depending on constructor argument.
At this point I'd call to one of options:
- move this to separate change request,
- go with size 500, as it's intended for single-email field (From); other option just doesn't sound right (using a string possibly contatining several emails in undefined order as a key)

Reply:
It's an interesting case. I see that it is used with both options in our code. Therefore, I would like to avoid hard-coding  an assumption that won't be always correct in the analysis. 

I would go with a separate CR with a normal priority which will be implemented in future versions. We have such CRs for other corner cases. If developers will complain about it, we can always increase the priority and add support earlier.